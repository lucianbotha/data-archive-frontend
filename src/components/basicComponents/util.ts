const updateState = (state: any, name: string, value: string) => {
  let newState = state;
  if (name === "targetName") {
    newState = {
      ...state,
      target: {
        ...state.target,
        targetName: {
          ...state.target.targetName,
          name: value
        }
      }
    };
  }
  if (name === "ra") {
    newState = {
      ...state,
      target: {
        ...state.target,
        ra: {
          ...state.target.ra,
          value
        }
      }
    };
  }
  if (name === "dec") {
    newState = {
      ...state,
      target: {
        ...state.target,
        dec: {
          ...state.target.dec,
          value
        }
      }
    };
  }
  if (name === "radius") {
    newState = {
      ...state,
      target: {
        ...state.target,
        radius: {
          ...state.target.radius,
          value
        }
      }
    };
  }
  if (name === "radiusUnits") {
    newState = {
      ...state,
      target: {
        ...state.target,
        radiusUnits: {
          ...state.target.radiusUnits,
          value
        }
      }
    };
  }
  if (name === "resolver") {
    newState = {
      ...state,
      target: {
        ...state.target,
        resolver: {
          ...state.target.resolver,
          value
        }
      }
    };
  }
  if (name === "proposalCode") {
    newState = {
      ...state,
      proposal: {
        ...state.proposal,
        proposalCode: {
          ...state.proposal.proposalCode,
          code: value
        }
      }
    };
  }
  if (name === "proposalTitle") {
    newState = {
      ...state,
      proposal: {
        ...state.proposal,
        proposalTitle: {
          ...state.proposal.proposalTitle,
          title: value
        }
      }
    };
  }
  if (name === "pi") {
    newState = {
      ...state,
      proposal: {
        ...state.proposal,
        pi: {
          ...state.proposal.pi,
          name: value
        }
      }
    };
  }
  if (name === "obsDate") {
    newState = {
      ...state,
      proposal: {
        ...state.proposal,
        obsDate: {
          ...state.proposal.obsDate,
          value
        }
      }
    };
  }
  if (name === "telescope") {
    newState = {
      ...state,
      telescope: {
        ...state.telescope,
        selectedTelescope: value
      }
    };
  }
  if (name === "instrument") {
    newState = {
      ...state,
      telescope: {
        ...state.telescope,
        selectedInstrument: value
      }
    };
  }
  if (name === "dataType") {
    newState = {
      ...state,
      data: {
        ...state.data,
        selectedDataType: value
      }
    };
  }
  if (name === "arcs") {
    console.log("arcs", value);
    newState = {
      ...state,
      data: {
        ...state.data,
        arcs: !state.data.arcs
      }
    };
  }
  if (name === "biases") {
    newState = {
      ...state,
      data: {
        ...state.data,
        biases: !state.data.biases
      }
    };
  }
  if (name === "standards") {
    newState = {
      ...state,
      data: {
        ...state.data,
        standards: !state.data.standards
      }
    };
  }
  if (name === "flats") {
    newState = {
      ...state,
      data: {
        ...state.data,
        flats: !state.data.flats
      }
    };
  }
  return newState;
};

const checkDateFormatError = (date: string) => {
  let error = "";
  const dateReg = /^[0-3]\d([./-])[0-1]\d\1\d{4}$/;
  if (date !== "" && !date.match(dateReg)) {
    error = "Date format does not match (date should be like [DD-MM-YYYY])";
  }
  return error;
};

const checkNameFormatError = (name: string) => {
  let error = "";
  const nameReg = /^[a-zA-Z0-9 ]*$/;
  if (name !== "" && !name.match(nameReg)) {
    error = "This should only contains characters contained in a name only";
  }
  return error;
};

const checkValueFormatError = (value: string) => {
  let error = "";
  const dateReg = /^[1-9]\d*(\.\d+)?$/;
  if (value !== "" && !value.match(dateReg)) {
    error = "This should be a number";
  }
  return error;
};

const validateQuery = (state: any) => {
  return {
    ...state,
    proposal: {
      obsDate: {
        ...state.proposal.obsDate,
        error: checkDateFormatError(state.proposal.obsDate.value)
      },
      pi: {
        ...state.proposal.pi,
        error: checkNameFormatError(state.proposal.pi.name)
      },
      proposalCode: {
        ...state.proposal.proposalCode,
        error: checkNameFormatError(state.proposal.proposalCode.code)
      },
      proposalTitle: {
        ...state.proposal.proposalTitle,
        error: checkNameFormatError(state.proposal.proposalTitle.title)
      }
    },
    target: {
      dec: {
        ...state.target.dec,
        error: checkValueFormatError(state.target.dec.value)
      },
      ra: {
        ...state.target.ra,
        error: checkValueFormatError(state.target.ra.value)
      },
      radius: {
        ...state.target.radius,
        error: checkValueFormatError(state.target.radius.value)
      },
      radiusUnits: {
        ...state.target.radiusUnits
      },
      resolver: {
        ...state.target.resolver
      },
      targetName: {
        ...state.target.targetName,
        error: checkNameFormatError(state.target.targetName.name)
      }
    }
  };
};
export { updateState, validateQuery };
