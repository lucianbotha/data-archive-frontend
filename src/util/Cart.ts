/**
 * The cart of requested files.
 */
export class Cart {
  constructor(private files: CartFile[]) {}

  /**
   * Check whether the cart contains the given file.
   *
   * Files are compared by their id. So if any cart file has the same id as the
   * given file, the function returns true, irrespective of any other key-value
   * pairs are equal.
   *
   * Parameters
   * ----------
   * file:
   *     The file whose existence in the cart is checked.
   *
   * Returns
   * -------
   * contains:
   *     Whether the cart contains the given file.
   */
  public contains(file: CartFile) {
    return this.files.some(f => f.id === file.id);
  }

  /**
   * Adds or removes files from the cart.
   *
   * A file in the given array is added if it is not in the cart yet, and it is
   * removed if it is in the cart already. The same logic as for the contains
   * method is used to decide whether a file is in the cart.
   *
   * @param files
   */
  public addOrRemove(files: CartFile[]) {
    files.forEach(file => {
      const cartFilesFilesWithoutFile = this.files.filter(
        f => f.id !== file.id
      );
      if (cartFilesFilesWithoutFile.length !== this.files.length) {
        this.files = [...cartFilesFilesWithoutFile];
      } else {
        this.files = [...this.files, file];
      }
    });
  }

  /**
   * Returns the cart files grouped by observation.
   *
   * The files are returned as a map whose keys are the observation ids. Each
   * value is the set of files for the observation which have the corresponding
   * key as their id.
   *
   * Files which are not linked to an observation are collected in a set with
   * the empty string as its key.
   *
   * For example, consider a cart containing the files
   *
   * A_Obs1, B, C_Obs1, D_Obs2, E
   *
   * where A_Obs1 and C_Obs1 are linked to observation with id Obs1, D_Obs2 is
   * linked to an observation with id Obs2 and B and E are not linked to any
   * observation. Then this method returns a map with the following content.
   *
   * {
   *   '': new Set(B, E),
   *   'Obs1': new Set(A_Obs1, C_Obs1),
   *   'Obs2': new Set(D_Obs2)
   * }
   */
  public groupByObservation() {
    const groups = new Map<string, Set<CartFile>>();
    this.files.forEach(file => {
      const key = (file.observation && file.observation.id) || "";
      if (!groups.has(key)) {
        groups.set(key, new Set<CartFile>());
      }
      (groups.get(key) as Set<CartFile>).add(file);
    });

    return groups;
  }
}

/**
 * Interface for a cart file.
 *
 * Properties
 * ----------
 * id:
 *     Unique file id.
 * name:
 *     File name.
 * observation:
 *     Observation to which the file is linked.
 */
export interface CartFile {
  id: string;
  name: string;
  observation?: Observation | null;
}

/**
 * Interface for an observation to which a cart file is linked.
 *
 * Properties
 * ----------
 * id:
 *     Unique observation id.
 * name:
 *     Observation name.
 */
interface Observation {
  id: string;
  name: string;
}
