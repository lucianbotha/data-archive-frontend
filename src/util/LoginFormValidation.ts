/**
 * An interface for the registration inputs
 *
 * Properties:
 * -----------
 * password:
 *     The password, which must have at least 7 characters.
 * username:
 *     The username, which must not contain upper case letters.
 */
export interface ILoginForm {
  password: string;
  username: string;
}

/**
 * Login form validation for the user input.
 *
 * It accepts an object argument as defined by the ILoginForm interface
 *
 * @param loginInput
 *
 * An example of the submitted user input may look like:
 *
 * loginInput = {
 *
 *    password: "SecureandMatchingPassword",
 *
 *    username: "validusername"
 * }
 *
 * It returns an erray of string items if there is any error message or
 * an empty array if there are no items.
 */

export const validateLoginField = (loginInput: ILoginForm) => {
  // An object to Store errors for all fields.
  const errors = {
    password: "",
    username: ""
  };

  // Check if the submitted username is not empty.
  if (!loginInput.username) {
    errors.username = `Username cannot be empty.`;
  }

  // Check if the submitted username contains upper case characters.
  if (loginInput.username !== loginInput.username.toLowerCase()) {
    errors.username = "Username must be in lowercase";
  }

  // Check if the password is secure enough.
  if (loginInput.password.length <= 6) {
    errors.password = "Password should be at least 7 characters long";
  }

  // return an array consisting of error message if any or empty.
  return errors;
};
