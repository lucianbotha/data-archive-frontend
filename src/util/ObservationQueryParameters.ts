// TypeScript interfaces describing the state of the search form

/**
 * An interface describing the React state of the observation query form.
 *
 * This state is used for populating all search form fields (and error messages)
 * and is parsed into the GraphQL query for searching observations.
 *
 * Properties:
 * -----------
 * general:
 *     General information about the observation.
 * target:
 *     Target details.
 * telescope:
 *     Telescope (and instrument) details.
 */
export interface IObservationQueryParameters {
  general?: IGeneral;
  target?: ITarget;
  telescope: ITelescope;
}

/**
 * An interface for query parameters related to general information.
 *
 * Properties:
 * -----------
 * errors:
 *     Error messages.
 * observationNight:
 *     Date of the night when the observation was taken. This
 *     refers to the date whe the night started.
 * proposalCode:
 *     Identifier of the proposal to which the observation belongs.
 * proposalTitle:
 *     Title of the proposal to which the observation belongs.
 *
 */
export interface IGeneral {
  errors: {
    observationNight?: string;
  };
  observationNight?: string;
  principalInvestigator?: string;
  proposalCode?: string;
}

/**
 * An interface for query parameters related to target details.
 *
 * Properties:
 * -----------
 * declination:
 *     Declination.
 * errors:
 *     Errors.
 * name:
 *     Target name.
 * resolver:
 *     Resolver to use for resolving the target name to a position.
 * rightAscension:
 *     Right ascension.
 * searchConeRadius:
 *     Radius of the search cone radius, in the units given by
 *     searchConeRadiusUnits.
 * searchConeRadiusUnits:
 *     Units of the search cone radius given by searchConeRadius.
 *
 */
export interface ITarget {
  declination?: string;
  errors: {
    declination?: string;
    rightAscension?: string;
    searchConeRadius?: string;
    searchConeRadiusUnits?: string;
  };
  name?: string;
  resolver?: string;
  rightAscension?: string;
  searchConeRadius?: string;
  searchConeRadiusUnits?: string;
}

// TELESCOPES

export type TelescopeName = "Lesedi" | "SALT" | "1.9 m";

/**
 * An interface for query parameters related to a telescope.
 *
 * This serves as a placeholder for more concrete interfaces such as ISALT.
 *
 * Properties:
 * -----------
 * name:
 *     Telescope name.
 */
export interface ITelescope {
  name: TelescopeName;
}

/**
 * An interface for query parameters related to SALT.
 *
 * Properties:
 * -----------
 * instrument:
 *     A SALT instrument.
 */
export interface ISALT extends ITelescope {
  name: "SALT";
  instrument?: IInstrument;
}

/**
 * An interface for query parameters related to Lesedi.
 *
 * Properties:
 * -----------
 * instrument:
 *     A SALT instrument.
 */
export interface ILesedi extends ITelescope {
  name: "Lesedi";
  instrument?: IInstrument;
}

/**
 * An interface for query parameters related to SALT.
 *
 * Properties:
 * -----------
 * instrument:
 *     A SALT instrument.
 */
export interface ILesedi extends ITelescope {
  name: "Lesedi";
  instrument?: IInstrument;
}

// INSTRUMENTS

/**
 * An interface for query parameters related to an instrument.
 *
 * This serves as a placeholder for more concrete interfaces such as ISalticam.
 *
 * Properties:
 * -----------
 * name:
 *     Telescope name.
 */
export interface IInstrument {
  name: string;
}

/**
 * An interface for query parameters related to Salticam.
 *
 * Properties:
 * -----------
 * detectorMode:
 *     Detector mode.
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 */
export interface ISalticam extends IInstrument {
  name: "Salticam";
  detectorMode?: "Normal" | "Slot Mode";
  errors: {
    detectorMode?: string;
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
}

/**
 * An interface for query parameters related to RSS.
 *
 * Properties:
 * -----------
 * detectorMode:
 *     Detector mode.
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 */
export interface IRSS extends IInstrument {
  name: "RSS";
  detectorMode?: "Normal" | "Slot Mode";
  errors: {
    detectorMode?: string;
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
}

/**
 * An interface for query parameters related to HRS.
 *
 * Properties:
 * -----------
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * mode:
 *     Resolution mode.
 */
export interface IHRS extends IInstrument {
  name: "HRS";
  errors: {
    exposureTime?: string;
    mode?: string;
  };
  exposureTime?: string;
  mode?:
    | "Low Resolution"
    | "Medium Resolution"
    | "High Resolution"
    | "High Stability";
}

/**
 * An interface for query parameters related to HIPPO.
 *
 * Properties:
 * -----------
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 */
export interface IHIPPO extends IInstrument {
  name: "HIPPO";
  errors: {
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
}

/**
 * An interface for query parameters related to SHOC.
 *
 * Properties:
 * -----------
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 */
export interface ISHOC extends IInstrument {
  name: "SHOC";
  errors: {
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
}
