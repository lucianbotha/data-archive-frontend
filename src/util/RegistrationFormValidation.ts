import { validate } from "isemail";
/**
 * An interface for the registration inputs
 *
 * Properties:
 * -----------
 * affiliation:
 *     The affiliation of the user, such as a university or an institute.
 * confirmPassword:
 *      A matching password that must match with the original password.
 * email:
 *     Email address. This will be stored in lower case.
 * familyName:
 *     The family name (surname).
 * givenName:
 *     The given name (first name).
 * password:
 *     The password, which must have at least 7 characters.
 * username:
 *     The username, which must not contain upper case letters.
 */
export interface IRegistrationForm {
  affiliation: string;
  confirmPassword: string;
  email: string;
  givenName: string;
  familyName: string;
  password: string;
  username: string;
}

/**
 * Registration form validation for the user input.
 *
 * It accepts an object argument as defined by the IRegistrationForm interface
 *
 * @param registrationInput
 *
 * An example of the submitted user input may look like:
 *
 * registrationInput = {
 *    affiliation: "University of Cape Town",
 *
 *    confirmPassword: "SecureandMatchingPassword",
 *
 *    emailAddress: "valid@gmail.com",
 *
 *    familyName: "Smith",
 *
 *    givenName: "John",
 *
 *    password: "SecureandMatchingPassword",
 *
 *    username: "validusername"
 * }
 *
 * It returns an erray of string items if the is any error message or
 * an empty array if there are no items.
 */

export const validateRegistrationField = (
  registrationInput: IRegistrationForm
) => {
  // An object to Store errors for all fields.
  const errors: string[] = [];

  // Check if the submitted given name is not empty.
  if (!registrationInput.givenName) {
    errors.push("Given name can't be empty");
  }

  // Check if the submitted family name is not empty.
  if (!registrationInput.familyName) {
    errors.push("Family name can't be empty");
  }

  // Check if the submitted username is not empty.
  if (!registrationInput.username) {
    errors.push(`Username cannot be empty.`);
  }

  // Check if the submitted username contains upper case characters.
  if (registrationInput.username !== registrationInput.username.toLowerCase()) {
    errors.push("Username must be in lowercase");
  }

  // Check if the submitted email address is valid.
  if (!validate(registrationInput.email, { minDomainAtoms: 2 })) {
    errors.push("Email address is invalid");
  }

  // Check if the password is secure enough.
  if (registrationInput.password.length <= 6) {
    errors.push("Password should be at least 7 characters long");
  }

  // Check if passwords match
  if (registrationInput.password !== registrationInput.confirmPassword) {
    errors.push("Passwords do not match");
  }

  // return an array consisting of error message if any or empty.
  return errors;
};
