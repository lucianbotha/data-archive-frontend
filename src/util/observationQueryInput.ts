import moment, { Moment } from "moment";
import { IGeneral } from "./ObservationQueryParameters";

const GENERAL_OBSERVATION_NIGHT = "A.ObsNight";

const GENERAL_PRINCIPAL_INVESTIGATOR = "A.PrincipalInvestigator";

const GENERAL_PROPOSAL_CODE = "A.ProposalCode";

export function generalWhereConstraints(general: IGeneral): any {
  const constraints = [];

  // Night when the observation was taken
  const observationNight = trim(general.observationNight);
  if (observationNight) {
    const startTime = parseDate(observationNight).add(14, "hours"); // noon SAST
    const startTimeString = startTime.format("YYYY-MM-DD");
    const endTime = startTime.add(1, "day");
    const endTimeString = endTime.format("YYYY-MM-DD");
    constraints.push(greaterThan(GENERAL_OBSERVATION_NIGHT, startTimeString));
    constraints.push(lessThan(GENERAL_OBSERVATION_NIGHT, endTimeString));
  }

  // Principal Investigator
  const principalInvestigator = trim(general.principalInvestigator);
  if (principalInvestigator) {
    constraints.push(
      contains(GENERAL_PRINCIPAL_INVESTIGATOR, principalInvestigator)
    );
  }

  // Proposal code
  const proposalCode = trim(general.proposalCode);
  if (proposalCode) {
    constraints.push(contains(GENERAL_PROPOSAL_CODE, proposalCode));
  }

  return constraints;
}

type DATABASE_VALUE = string | number | boolean;

function equals(column: string, value: DATABASE_VALUE) {
  return {
    EQUALS: { column, value }
  };
}

function lessThan(column: string, value: DATABASE_VALUE) {
  return {
    LESS_THAN: { column, value }
  };
}

function greaterThan(column: string, value: DATABASE_VALUE) {
  return {
    GREATER_THAN: { column, value }
  };
}

function contains(column: string, value: DATABASE_VALUE) {
  return {
    CONTAINS: { column, value }
  };
}

function trim(value: string | null | undefined) {
  if (value) {
    return value.trim();
  }
  return value;
}

/**
 * Parse the given string as a date, and return the date as a string as a
 * Moment instance.
 *
 * Thec string must have the format 'YYYY-MM-DD', and it must be a valid date.
 *
 * The date is assumed to be given as UTC.
 */
function parseDate(date: string) {
  const t = moment.utc(date, "YYYY-MM-DD");
  if (!t.isValid()) {
    throw new Error(`${date} is not a valid date.`);
  }
  t.startOf("day");
  return t;
}
