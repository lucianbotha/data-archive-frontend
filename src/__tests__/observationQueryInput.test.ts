import { generalWhereConstraints } from "../util/observationQueryInput";

describe("generalWhereConstraint", () => {
  const useInvalidObservationNightFormat = () => {
    generalWhereConstraints({ errors: {}, observationNight: "invaliddate" });
  };

  const useInvalidObservationNight = () => {
    generalWhereConstraints({ errors: {}, observationNight: "2019-02-29" });
  };

  it("should map an observation night to a greater than and less than constraint spanning one night", () => {
    expect(
      generalWhereConstraints({ errors: {}, observationNight: "2019-02-17" })
    ).toMatchSnapshot();
  });

  it("should raise an error for an invalid observation night format", () => {
    expect(useInvalidObservationNightFormat).toThrow(/invaliddate.*valid/);
  });

  it("should raise an error for an invalid observation night", () => {
    expect(useInvalidObservationNight).toThrow(/2019-02-29.*valid/);
  });

  it("should map a Principal Investigator to a contains constraint", () => {
    expect(
      generalWhereConstraints({ errors: {}, principalInvestigator: "Doe" })
    ).toMatchSnapshot();
  });

  it("should map a proposal code to a contains constraint", () => {
    expect(
      generalWhereConstraints({ errors: {}, proposalCode: "2-SCI" })
    ).toMatchSnapshot();
  });

  it("should map input data correctly", () => {
    expect(
      generalWhereConstraints({
        errors: {},
        observationNight: "2019-05-13",
        principalInvestigator: "Sipho",
        proposalCode: "2019-1-SCI-042"
      })
    ).toMatchSnapshot();
  });
});

describe("parseTargetPosition", () => {});
