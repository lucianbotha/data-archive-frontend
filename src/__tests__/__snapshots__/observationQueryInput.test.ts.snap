// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generalWhereConstraint should map a Principal Investigator to a contains constraint 1`] = `
Array [
  Object {
    "CONTAINS": Object {
      "column": "A.PrincipalInvestigator",
      "value": "Doe",
    },
  },
]
`;

exports[`generalWhereConstraint should map a proposal code to a contains constraint 1`] = `
Array [
  Object {
    "CONTAINS": Object {
      "column": "A.ProposalCode",
      "value": "2-SCI",
    },
  },
]
`;

exports[`generalWhereConstraint should map an observation night to a greater than and less than constraint spanning one night 1`] = `
Array [
  Object {
    "GREATER_THAN": Object {
      "column": "A.ObsNight",
      "value": "2019-02-17",
    },
  },
  Object {
    "LESS_THAN": Object {
      "column": "A.ObsNight",
      "value": "2019-02-18",
    },
  },
]
`;

exports[`generalWhereConstraint should map input data correctly 1`] = `
Array [
  Object {
    "GREATER_THAN": Object {
      "column": "A.ObsNight",
      "value": "2019-05-13",
    },
  },
  Object {
    "LESS_THAN": Object {
      "column": "A.ObsNight",
      "value": "2019-05-14",
    },
  },
  Object {
    "CONTAINS": Object {
      "column": "A.PrincipalInvestigator",
      "value": "Sipho",
    },
  },
  Object {
    "CONTAINS": Object {
      "column": "A.ProposalCode",
      "value": "2019-1-SCI-042",
    },
  },
]
`;
